### ISSUE TITLE
feat(autodiff): LBFGS optimizer（Hessian 近似）の実装

### ISSUE THREAD

**Issue by aLiz-Nancy** [Fandhe-AI/fandhe-ai#2172]
## 概要

LBFGS（Limited-memory BFGS）optimizer の実装。他の 1 階 gradient-based optimizer と異なり closure を複数回評価して line search を行うため、実装・テストの粒度が大きい。実装の複雑度により 2 sub に分割：(a) 値型実装・(b) compile 統合・テスト。

根拠: #2171（Adadelta・Adamax・NAdam・RAdam 実装）と同じ値型・純関数方式の延長。PyTorch `torch.optim.LBFGS`。

## 対象範囲

- `crates/autodiff/src/nn/optim/lbfgs.rs`（新規）：値型・closure メカニズム・line search
- `crates/facade/src/compat/training.rs`：`Optimizer` enum への `Lbfgs` variant 追加
- `crates/facade/tests/compat_sequential_*.rs`：LBFGS で compile・fit のテスト

## 受け入れ条件

- [ ] 分割方針: 実装を下記 2 sub へ分割
  - #2197：値型・closure 梯度評価・line search アルゴリズム
  - #2198：compile enum・fit 経由テスト・全 optimizer との互換性
- [ ] 値型実装が PyTorch LBFGS の 1 step と等価の収束を確認
- [ ] fit（MNIST 規模）での learning curve が SGD 相当同等以上の改善を示す

## 契約

- tolerance／baseline／`Cargo.toml` 依存／ガードレール閾値／`docs/spec/`（正本）は変更しない。仕様変更は spec リポへの (b) 形式提案としてユーザー承認後に行う
- ユーザー承認事項（依存追加・新規 `unsafe`・facade 公開面の拡張・spec 提案の投稿）は本文に列挙し、承認前に実施しない
- CUDA／Metal 実機 parity は `#[ignore]` で分離し、未実測なら `docs/perf/logs/<slug>-<issue 番号>/` へ申し送る
- crates.io 出荷済み `fandhe-ai =0.9.0` の公開 API を破壊しない（既存メソッドのシグネチャ・意味論は不変。拡張は追加 API・追加フィールド・opt-in フラグ・`#[non_exhaustive]` enum の variant 追加に限る）
- 新規演算は CPU 参照実装を先行し、CUDA／Metal は既定 `Unsupported` フォールバック（ホスト計算）で到達可能にする。GPU 専用カーネルは本 issue の対象外（別 issue）

## 承認事項

- facade 公開面の拡張（`docs/compat-api-scope.md` §5 経路 2）：`Optimizer::Lbfgs` variant・`LbfgsConfig` 型

## 分割方針

本 issue は実装の複雑度により下記 2 sub に分割：
- **#2197** `feat(autodiff): LBFGS 値型実装`：値型・closure 梯度評価・line search アルゴリズム
- **#2198** `feat(facade): compile() への LBFGS 統合とテスト`：compile enum・fit 経由テスト

## スコープ外

- 並列化（複数 worker での line search 並行実行）
- second-order method（牛頓法等）

## 参照

- #2171（Adadelta・Adamax・NAdam・RAdam）
- PyTorch `torch.optim.LBFGS` documentation
- `docs/compat-api-scope.md` §1.3（optimizer 網羅）
